New Demo Mode

jankatins opened the pull request:
Builds new notebook cells from the source of a function:

Example:

```
from ipyext.demo import demo
import ipyext.demo
# ---
demo(ipyext.demo)
# ---
demo(ipyext.demo.demo_example)
```

Closes: https://github.com/ipython-contrib/IPython-extensions/issues/13

**Todo**
- [x] write some tests for the ipython frontend
- [x] `demo("<gh_matplotlib>/")` -> prints the wrong thing...
- [x] write docs...
- [ ] How to format github examples to get them split into multiple cells? Look at some more matplotlib examples...
- [x] Use the events: `get_ipython().events.register("post_run_cell", self.method)`, see here:  https://github.com/FrancescAlted/ipython_memwatcher/blob/master/ipython_memwatcher/memwatcher.py#L47

jankatins committed:
New Demo Mode

This will either use a special module/function or a special URL at github as
the source for the next cells in a notebook.

Currently only the notebook version is available.

```
from ipyext.demo import notebook_demo
import ipyext.demo
notebook_demo(ipyext.demo)
notebook_demo(ipyext.demo.demo_example)
notebook_demo("<gh_matplotlib>/animation/")
notebook_demo("<gh_matplotlib>/animation/animate_decay.py")
```

jankatins committed:
Make demo work in ipython/qtconsole

This is a bit of a hack: the run_cell method is wrapped so that as
long as there is demo materialin a buffer, after the execution of the
cell, the next cell is set to the next demo material.

jankatins commented:
And now I have it working in the notebook and in ipython/qtconsole directly...

Also added matplotlib examples directly from github...

juhasch commented:
Nice!
Seems like `notebook_demo` is actutally `demo` ?

``` Python
import ipyext.demo
from ipyext.demo import demo
demo(ipyext.demo.demo_example)
```

jankatins commented:
@juhasch yep... Renamed it after I got the ipython version running

jankatins commented:
Example and some demo of the demo mode: http://janschulz.github.io/demo_mode_ipython.html

juhasch commented:
Looks like it could be really useful.

jankatins commented:
Lets see if someone uses it... Having the function here and not in ipython is a little entry barrier :-/

jankatins committed:
Doc: add documentation about demo mode

jankatins committed:
Demo: nicer TOC format for github entries

jankatins committed:
Demo: refactor source splitting

jankatins committed:
Demo: use a event handler instead of wrapping a internal method

jankatins committed:
Demo: make demo('STOP') work

jankatins committed:
Demo: fix github ToC on py2

name was used multiple times as variable: once for the name of the
current content and several times in list comprehensions. On py2,
this results in the name variable overwritten...

jankatins committed:
Demo: new tests for the frontends

jankatins committed:
Demo: more documentation

jankatins commented:
Ok, IMO this is ready if travis gets green...

The only open todo is how to handle matplotlib examples, but I tested this and the current approach (aka not splitting it into multiple cells) is the only one which works due to the way they are written (comments with `#` and also because they expect that `plt.show()` is called in the end... It's even worse when you use `%matplotlib inline` as that produces one plot per cell :-()

If it is merged I will write demos for the current magics and then lets see if there are some other projects which want to write demos...

jankatins commented:
@juhasch a short review and if satisfied a merge would be nice :-)

juhasch commented:
Hmm,
tried to run tests and generate docs:

```
$ make
python autogen_api.py
python: can't open file 'autogen_api.py': [Errno 2] No such file or directory
make: *** [source/api/generated/gen.txt] Error 2

$ iptest3 ipyext

Test group: ipyext
E
======================================================================
ERROR: Failure: ImportError (No module named 'ipyext')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/users/juhasch/anaconda/lib/python3.5/site-packages/nose/failure.py", line 39, in runTest
    raise self.exc_val.with_traceback(self.tb)
  File "/users/juhasch/anaconda/lib/python3.5/site-packages/nose/loader.py", line 407, in loadTestsFromName
    module = resolve_name(addr.module)
  File "/users/juhasch/anaconda/lib/python3.5/site-packages/nose/util.py", line 312, in resolve_name
    module = __import__('.'.join(parts_copy))
ImportError: No module named 'ipyext'

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (errors=1)

______________________________________________________________________
Test suite completed for system with the following information:
IPython version: 4.0.0
IPython commit : f534027 (installation)
IPython package: /users/juhasch/anaconda/lib/python3.5/site-packages/IPython
Python version : 3.5.0 |Anaconda 2.4.0 (x86_64)| (default, Oct 20 2015, 14:39:26) [GCC 4.2.1 (Apple Inc. build 5577)]
sys.executable : /users/juhasch/anaconda/bin/python
Platform       : Darwin-15.2.0-x86_64-i386-64bit

Tools and libraries available at test time:
   matplotlib pygments sqlite3

Status: ERROR - 1 out of 1 test groups failed (ipyext). Took 1.191s.

You may wish to rerun these, with:
  iptest ipyext
```

jankatins committed:
Doc: newer makefile

jankatins commented:
The first (make) should be fixed (I use make.cmd and RTD something else, so it wasn't catched,,,)

jankatins commented:
I can't reproduce the second problem. Could it be that you ran the tests in the docs subdir?

juhasch commented:
Adding the extension to `PYTHONPATH` let's me run the tests, but now they loop endlessly.
Will look into it tomorrow.

The doc is now generated fine.

jankatins commented:
there is also `python tests.py` but that should actualy be mostly the same as `iptest ipyext`.

jankatins committed:
Add seaborn as a new example source

jankatins committed:
Demo: whatsnew entry

juhasch commented on ipyext/demo.py:
```diff
@@ -0,0 +1,533 @@
+    print(_sum)
+
+# This lets the `demo(ipyext.demo)` find only the `demo_example`.
+# Only modues with that variable will display an overview of
```
Type: modues->modules

juhasch commented:
Works fine in the notebook.
In the shell I get

``` Python
In [4]: demo(ipyext.demo.demo_example)
<IPython.core.display.Javascript object>
```

Would it make sense to return a text representation in this case ?

jankatins commented:
> In [4]: demo(ipyext.demo.demo_example)
> IPython.core.display.Javascript object
> Would it make sense to return a text representation in this case ?

You need to use `demo(..., frontend="ipython")`... I've actually no idea how to find out if a kernel is run in a notebook or in the qtconsole/console :-(

jankatins committed:
Demo: send a explaining message if wrong frontend is used

Curently I don't know any way to detect which frontent is used to
run a kernel, so if the demo is used in the defualt mode (=notebook)
make it at least output a helpful message if the frontend does not
understand the javscript code to insert the new cells.

Alos fix some typos...

juhasch commented:
How about adding a `__repr__()` function ?
This should be called if `_repr_javascript_` is not supported at the output.

jankatins commented:
Currently I use the normal `JavaScript` object... Hm, I would need to subclass that... ok, good idea, incoming... :-)

jankatins commented:
done :-)
